pkg/user/manager/rest/rest.go

```diff
@@ -0,0 +1,341 @@
+// Copyright 2018-2020 CERN
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// In applying this license, CERN does not waive the privileges and immunities
+// granted to it by virtue of its status as an Intergovernmental Organization
+// or submit itself to any jurisdiction.
+
+package rest
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net/url"
+	"strings"
+	"time"
+
+	"github.com/cs3org/reva/pkg/rhttp"
+	"github.com/cs3org/reva/pkg/user"
+	"github.com/cs3org/reva/pkg/user/manager/registry"
+	"github.com/mitchellh/mapstructure"
+
+	userpb "github.com/cs3org/go-cs3apis/cs3/identity/user/v1beta1"
+)
+
+func init() {
+	registry.Register("rest", New)
+}
+
+type manager struct {
+	conf                *config
+	apiToken            string
+	tokenExpirationTime time.Time
+}
+
+type config struct {
+	APIBaseURL   string `mapstructure:"api_base_url"`
+	ClientID     string `mapstructure:"client_id"`
+	ClientSecret string `mapstructure:"client_secret"`
+
+	OIDCTokenEndpoint string `mapstructure:"oidc_token_endpoint"`
+	TargetAPI         string `mapstructure:"target_api"`
+
+	IDProvider string `mapstructure:"id_provider"`
+}
+
+func parseConfig(m map[string]interface{}) (*config, error) {
+	c := &config{}
+	if err := mapstructure.Decode(m, c); err != nil {
+		return nil, err
+	}
+	return c, nil
+}
+
+// New returns a user manager implementation that makes calls to the GRAPPA API.
+func New(m map[string]interface{}) (user.Manager, error) {
+	c, err := parseConfig(m)
+	if err != nil {
+		return nil, err
+	}
+
+	if c.APIBaseURL == "" {
+		c.APIBaseURL = "https://authorization-service-api-dev.web.cern.ch/api/v1.0"
+	}
+
+	if c.TargetAPI == "" {
+		c.TargetAPI = "authorization-service-api"
+	}
+
+	if c.OIDCTokenEndpoint == "" {
+		c.OIDCTokenEndpoint = "https://keycloak-dev.cern.ch/auth/realms/cern/api-access/token"
+	}
+
+	if c.IDProvider == "" {
+		c.IDProvider = "http://cernbox.cern.ch"
+	}
+
+	return &manager{
+		conf: c,
+	}, nil
+}
+
+func (m *manager) renewAPIToken(ctx context.Context) error {
```

Comment:
if this call can occur concurrently, we need to protect the m.apiToken and m.tokenExpirationTime with locks.

Reply:
Done